### DISCUSSION TITLE
[Modules Support] assign Modules to Hosts

### DISCUSSION THREAD

**Discussion by Spiegie** [linkdd/tricorder#25] [Ideas]
I'm working on Modules (see https://github.com/linkdd/tricorder/pull/24)

While working on it, I got some design-problems. 

Currently, you can call 
`./tricorder -i <inventory> module -- mod`, which is very manual.

I already implemented the call 
`./tricorder -i <inventory> module --all`
which aims to make the command above a bit more flexible. 

But here some options emerge. 
How should we assign modules to hosts in the inventory?

My first thought was, that there is a File, that assigns modules to a list of hosts (like saltstack is doing it with its top.sls).
You could combine this by assigning modules to certain tags.
My last thought was, that you should be able to pass a list of modules with the Inventory-file.

Maybe all three possibilities should exist.
We need to make sure, the Modules don't get executed multiple times.

I want to implement as many Ideas as I can, so we can play with them a bit, before we decide. 

What do you think? Do you have additional ideas? Is there an idea, you can cross out in advance?

**Comment by linkdd** [comment]
Ok, thanks for the clarifications.

In the PR, I missed the part where you wanted to match modules with hosts via tags. The structure of a `tricorder` command is as follow:

```
$ tricorder -i /path/to/inventory <target hosts> <task>
```

The `<target hosts>` part can be:

 - explicit: `-H <host name>`
 - by tags: `-t <tag selection>`

Do we really need another way of selecting hosts that will be specific to the module task?

What I understood from your idea of modules was something similar to the "external" tasks. Currently, `tricorder` provides builtin tasks (info, download, upload, exec). If you try to call `tricorder` with an unknown task, it will try to run the binary `tricorder-<the task>`. But that binary is executed locally.

It was a nice way to create "task bundles", for example, we could have a script called `tricorder-module` which does:

```bash
#!/bin/sh

set -e

tricorder -i $TRICORDER_INVENTORY -t $TRICORDER_HOST_TAGS upload /local/path/to/the-binary /remote/path/to/the-binary
tricorder -i $TRICORDER_INVENTORY -t $TRICORDER_HOST_TAGS do -- /remote/path/to/the-binary
```

Now, I agree that having this feature as a native task, which will handle the upload of the binary to a predetermined directory (`$HOME/.local/tricorder/bin/` ?), could be nice.

But reading your explanation now, I'm not sure I completely understood what you meant by "modules". For example, what is this command is supposed to do?

```
$ tricorder -i <inventory> module --all
```

The `--all` part is what's confusing me.

**Comment by Spiegie** [comment]
Thanks for your reply, and I'm sorry that I am a bit confusing :) I will try to make more clear, what I'm trying to accomplish 

When beginning my work on the modules, I tried to make arbitrary modules, that can be fed with plain data. These Modules then sit on many different hosts. The hosts then applies these modules. 

The data, that gets fed to the modules sits in the host variables (I think we will need an additional way of feeding the data, because the host variables could get very big very fast and could use some reusability, but that should be an own discussion).

So your understanding is very on point. 

By the way, Initially I tried to work out, how the host than can hold the state, we were giving it by running the module command. That's why, I wanted to upload the Module. But I dropped that thought today and returned to a more simple approach.

With the `--all` flag, I wanted to have a possibility, to run all modules, that I have local (in the data directory of the modules supplied by the `--data` flag). I did not want to apply it to all hosts, but to all hosts in the inventory, that should run certain modules. 

The use case is as follows:

A Server-Environment has many Mariadb Databases, but also some others servers, like a Grafana-Server and two Webserver in high-availability-mode. Now I want to configure all Mariadb servers very similar, but not equal. All hosts share a certain Network-configuration. The Grafana-Host and the Webserver share a certain group of users. All these Configurations can be rolled out by modules. But I don't want to call a tricorder-command for each module and specify a tag (which could go wrong). Furthermode, I think the list of modules could grow very big.

The `--all` flag signals, that I want to roll out all modules. Now I need to know, which hosts get which modules. That's why, I think, we need an additional way apart from the target host approach. At this point, I just want you to know, that I will fully accept it, if you don't like the Idea.

I hope, that it's clear now, what I want to accomplish. I will try to make it even more clear by continuing my work on the pull request. 

By the way, I have another idea. We could define a new object called "roles". Roles could match for a specific tag and link certain modules to certain hosts in the inventory.

**Comment by linkdd** [comment]
> I think we will need an additional way of feeding the data, because the host variables could get very big very fast and could use some reusability, but that should be an own discussion

If you know ansible, you may recall hosts, hostgroups, host vars and group vars. In ansible, every host is part of one or more group. Every host has variables, and every group has variables.

Here, instead of hostgroups, I have tags (but they are technically the same). The only thing I'm missing, is group vars.

Ansible provided, on top of the inventory file, the folders "group_vars" and "host_vars". I decided that this is un-needed and that if you want more complex inventories (and re-usability), you should use dynamic inventories.

In almost every use case I had professionally with Ansible, we were implementing dynamic inventories anyway, pulling the data from CMDBs and other third-party tools.

---

> With the `--all` flag, I wanted to have a possibility, to run all modules

Ok, got it.

---

> I did not want to apply it to all hosts, but to all hosts in the inventory, that should run certain modules.

Seems like you want a script :)

`tricorder` is a building-block that can be embedded in a shell script, or via the Rust API (and bindings to other languages were planned as well) in a real programming environment (no YAML programming, that's a non-goal).

---

Here is an idea, I'd like your feedback:

```
$ tricorder -i /path/to/inventory -H some_host module --bin /path/to/module.exe --data /path/to/data.json
```

Then, the `module::Task` would be implemented like this:

 - the `prepare()` function would read the data (eventually inject host vars into it)
 - the `apply()` function would upload the `module.exe` to the host, then execute the command `module.exe`, and write the data read in the previous method to the process's `stdin`

Something like this (proper error handling omitted):

```rust
channel.exec("/path/to/module.exe").unwrap();

let data = "the data read in prepare()";
channel.write_all(data.as_bytes()).unwrap();
channel.send_eof().unwrap();

let mut stdout = String::new();
channel.read_to_string(&mut stdout).unwrap();
let mut stderr = String::new();
channel.stderr().read_to_string(&mut stderr).unwrap();

channel.wait_close().unwrap();
let exit_code = channel.exit_status().unwrap();
```

This way, the `module` task is able to upload a binary (which will expect to read its data from stdin), execute it, and return the result (stdout/stderr).

This is a building block.

To provide the `--all` feature which somehow knows what host to target for which module, you'd need to either:

 - create a dynamic inventory, and a shell script which call the different `tricorder` commands
 - use the [external](https://docs.rs/tricorder/latest/tricorder/cli/external/index.html) task to bundle multiple commands together, and the [info](https://docs.rs/tricorder/latest/tricorder/tasks/info/index.html) task (with `jq`) to dynamically determine which module to run
 - use the Rust API (or any future bindings for other languages) to do all that in a real programming environment

This was the initial goal of the project (providing building blocks), but I'm open to criticism and other ideas.

What do you think?

**Comment by Spiegie** [comment]
I think, that I'm still thinking too complex. Thank you, for your suggestions. 

Using this approach, the user can make even better linking of modules with the hosts if he needs to and the library stays very simple. 

By the way, thanks for your suggestion, that the data should be read from stdin. I implemented it right away. I was stuck on providing the data by argument, and it was not very nice.
